File: tt_bio/kernels/triatt_bw/compute/triatt_bw.cpp
// SPDX-License-Identifier: Apache-2.0
//
// The triangle-attention backward, with the score block never leaving L1.
//
// One core owns one head and a contiguous group of the leading axis. For each row of that group it
// forms S = (Q K^T) * scale + bias, softmaxes it, and computes dV, dP, dS, dQ and dK without ever
// writing a score-sized tensor to DRAM. dbias is the only term that reduces across the leading
// axis, so it accumulates in float32 in this core's own L1 across the whole group and is written
// once at the end; a host-side sum over the group axis finishes it.
//
// The key axis is never chunked, which is why there is no log-sum-exp bookkeeping here: this core
// holds every key for its rows, so each softmax is exact and complete the first time. This file is
// the whole-query form, where one query chunk covers the token axis and dV and dK therefore need
// no accumulation across chunks. The host gate refuses any shape that does not fit that.
//
// The convention matched is autograd.triangle_attention's -- softmax(Q K^T * scale + bias) -- and
// NOT the fused forward's softmax((Q K^T + mask) * scale). The two differ by where the scale lands
// relative to the bias, and getting it backwards is a wrong gradient that the forward agrees with.

#define REDUCE_OP (PoolType::MAX)
#define REDUCE_DIM (ReduceDim::REDUCE_ROW)

#include <cstdint>

#include "api/compute/compute_kernel_api.h"
#include "api/compute/transpose_wh.h"
#include "../../triatt_sdpa/compute/compute_common.hpp"

namespace {

// out_cb[M, N] = in0_cb[M, K] @ in1_cb[K, N], and NEITHER input is popped.
//
// compute_common's matmul_blocks pops its in1, which does not work here: k is read once as K^T
// (for S) and once as K (for dQ) off the same tiles, and q is read once for S and once for dK.
// TRANSPOSE transposes in1's faces, which is what turns the Nt tiles of K into the tile grid of
// K^T -- for a [Nt, 1] operand the tile ORDER is already the transpose, so only the faces move.
template <uint32_t M, uint32_t N, uint32_t K, uint32_t SBH, uint32_t SBW, bool TRANSPOSE>
ALWI void mm_keep(uint32_t in0_cb, uint32_t in1_cb, uint32_t out_cb) {
    mm_block_init_short(in0_cb, in1_cb, TRANSPOSE, SBW /*ct_dim*/, SBH /*rt_dim*/, K /*kt_dim*/);
    reconfig_data_format(in1_cb, in0_cb);
    cb_wait_front(in0_cb, M * K);
    cb_wait_front(in1_cb, K * N);
    cb_reserve_back(out_cb, M * N);
    for (uint32_t m = 0; m < M; m += SBH) {
        for (uint32_t n = 0; n < N; n += SBW) {
            tile_regs_acquire();
            uint32_t in0_index = m * K;
            uint32_t in1_index = n;
            for (uint32_t kk = 0; kk < K; ++kk) {
                matmul_block(in0_cb, in1_cb, in0_index, in1_index, 0 /*dst*/, TRANSPOSE, SBW, SBH, K);
                in0_index++;
                in1_index += N;
            }
            tile_regs_commit();
            tile_regs_wait();
            uint32_t dst_idx = 0;
            for (uint32_t r = 0; r < SBH; ++r) {
                for (uint32_t c = 0; c < SBW; ++c) {
                    pack_tile<true>(dst_idx++, out_cb, (m + r) * N + n + c);
                }
            }
            tile_regs_release();
        }
    }
    cb_push_back(out_cb, M * N);
}

// out_cb[C, R] = transpose(in_cb[R, C]), both the tile grid and each tile's own faces.
//
// dV needs P^T and dK needs dS^T, and matmul_block can only transpose its in1. Doing it here costs
// R*C tile operations inside L1 and no DRAM at all, which is the trade this whole kernel is about.
template <uint32_t R, uint32_t C>
ALWI void transpose_block(uint32_t in_cb, uint32_t out_cb) {
    transpose_wh_init_short(in_cb);
    reconfig_data_format_srca(in_cb);
    pack_reconfig_data_format(out_cb);
    cb_wait_front(in_cb, R * C);
    cb_reserve_back(out_cb, R * C);
    for (uint32_t i = 0; i < C; ++i) {
        for (uint32_t j = 0; j < R; ++j) {
            acquire_dst();
            transpose_wh_tile(in_cb, j * C + i, 0);
            pack_tile<true>(0, out_cb, i * R + j);
            release_dst();
        }
    }
    cb_push_back(out_cb, R * C);
}

// out_cb = in0_cb * in1_cb, with NEITHER input popped.
// compute_common only has the in-place form, and dS needs P kept: the row-sum correction reads
// dP * P and then dS itself is P * (dP - correction), so P is live across both.
ALWI void mul_block_to(uint32_t in0_cb, uint32_t in1_cb, uint32_t out_cb, uint32_t num_tiles) {
    mul_tiles_init(in0_cb, in1_cb);
    cb_wait_front(in0_cb, num_tiles);
    cb_wait_front(in1_cb, num_tiles);
    cb_reserve_back(out_cb, num_tiles);
    for (uint32_t i = 0; i < num_tiles; ++i) {
        acquire_dst();
        mul_tiles(in0_cb, in1_cb, i, i, 0);
        pack_tile<true>(0, out_cb, i);
        release_dst();
    }
    cb_push_back(out_cb, num_tiles);
}

// in0_cb -= in1_cb broadcast along columns, in place. compute_common has the multiply form and the
// subtract-then-exp form, but not the plain subtract the softmax backward needs.
template <uint32_t rows, uint32_t cols>
ALWI void sub_block_bcast_cols_inplace(uint32_t in0_cb, uint32_t in1_cb) {
    sub_bcast_cols_init_short(in0_cb, in1_cb);
    cb_wait_front(in0_cb, rows * cols);
    cb_wait_front(in1_cb, rows);
    for (uint32_t i = 0; i < rows; ++i) {
        for (uint32_t j = 0; j < cols; ++j) {
            acquire_dst();
            sub_tiles_bcast_cols(in0_cb, in1_cb, i * cols + j, i, 0);
            pack_tile<true>(0, in0_cb, i * cols + j);
            release_dst();
        }
    }
}

// Fill num_tiles of cb with copies of zero_cb's single tile, and push them.
// The dbias accumulator has to start at zero and stay front-resident for the whole group, so it is
// produced once here and then only ever added into in place.
ALWI void seed_zeros(uint32_t cb, uint32_t zero_cb, uint32_t num_tiles) {
    copy_tile_to_dst_init_short(zero_cb);
    reconfig_data_format_srca(zero_cb);
    pack_reconfig_data_format(cb);
    cb_wait_front(zero_cb, 1);
    cb_reserve_back(cb, num_tiles);
    for (uint32_t i = 0; i < num_tiles; ++i) {
        acquire_dst();
        copy_tile(zero_cb, 0, 0);
        pack_tile<true>(0, cb, i);
        release_dst();
    }
    cb_push_back(cb, num_tiles);
}

}  // namespace

void kernel_main() {
    constexpr uint32_t Nt = get_compile_time_arg_val(0);          // token axis, in tiles
    constexpr uint32_t Dt = get_compile_time_arg_val(1);          // head dim, in tiles
    constexpr uint32_t H = get_compile_time_arg_val(2);
    constexpr uint32_t scale_fp32 = get_compile_time_arg_val(3);  // head_dim ** -0.5
    constexpr uint32_t sq_sbh = get_compile_time_arg_val(4);      // subblock for a [Nt, Nt] result
    constexpr uint32_t sq_sbw = get_compile_time_arg_val(5);
    constexpr uint32_t col_sbh = get_compile_time_arg_val(6);     // subblock for a [Nt, Dt] result
    constexpr uint32_t col_sbw = get_compile_time_arg_val(7);

    const uint32_t row_start = get_arg_val<uint32_t>(0);          // this core's leading-axis group
    const uint32_t row_end = get_arg_val<uint32_t>(1);

    constexpr uint32_t cb_q = tt::CBIndex::c_0;
    constexpr uint32_t cb_k = tt::CBIndex::c_1;
    constexpr uint32_t cb_v = tt::CBIndex::c_2;
    constexpr uint32_t cb_do = tt::CBIndex::c_3;
    constexpr uint32_t cb_bias = tt::CBIndex::c_4;
    constexpr uint32_t cb_scalar = tt::CBIndex::c_5;   // packed bf16 1.0, the reductions' scale
    constexpr uint32_t cb_zero = tt::CBIndex::c_6;
    constexpr uint32_t cb_scale = tt::CBIndex::c_7;    // the attention scale, as one bf16 tile
    constexpr uint32_t cb_p = tt::CBIndex::c_24;
    constexpr uint32_t cb_dp = tt::CBIndex::c_25;
    constexpr uint32_t cb_t = tt::CBIndex::c_26;
    constexpr uint32_t cb_row_a = tt::CBIndex::c_27;
    constexpr uint32_t cb_row_b = tt::CBIndex::c_28;
    constexpr uint32_t cb_dbias = tt::CBIndex::c_29;
    constexpr uint32_t cb_done = tt::CBIndex::c_30;
    constexpr uint32_t cb_dq = tt::CBIndex::c_16;
    constexpr uint32_t cb_dk = tt::CBIndex::c_17;
    constexpr uint32_t cb_dv = tt::CBIndex::c_18;

    constexpr uint32_t score_tiles = Nt * Nt;
    constexpr uint32_t col_tiles = Nt * Dt;

    mm_init(cb_q, cb_k, cb_p);

    // The bias is [1, H, N, N] broadcast over the leading axis, so this core's head slice is read
    // once by the reader and indexed for every row of the group rather than re-read per row. That
    // is the same argument the forward's persistent-mask CB makes, and it is worth more here
    // because the backward touches the bias twice.
    cb_wait_front(cb_bias, score_tiles);
    cb_wait_front(cb_scalar, 1);
    cb_wait_front(cb_scale, 1);

    seed_zeros(cb_dbias, cb_zero, score_tiles);

    for (uint32_t row = row_start; row < row_end; ++row) {
        // ---- S = (Q K^T) * scale + bias, then P = softmax(S) --------------------------------
        // cb_k holds K's Nt tiles. Read as [Dt, Nt] with the faces transposed, those same tiles
        // are K^T's tile grid, which is why no second copy of k is read from DRAM.
        mm_keep<Nt, Nt, Dt, sq_sbh, sq_sbw, true>(cb_q, cb_k, cb_p);
        mul_block_bcast_scalar_inplace<cb_scale, score_tiles>(cb_p);
        add_block_inplace<false>(cb_p, cb_bias, score_tiles);

        reduce_c<PoolType::MAX, ReduceDim::REDUCE_ROW, cb_p, cb_scalar, Nt, (int)VectorMode::RC>(
            cb_row_a, cb_row_a, Nt, false);
        sub_exp_block_bcast_cols_inplace<cb_p, Nt, 0x3F800000 /*1.0f*/>(cb_row_a, cb_row_b, Nt);
        cb_pop_front(cb_row_a, Nt);
        reduce_c<PoolType::SUM, ReduceDim::REDUCE_ROW, cb_p, cb_scalar, Nt, (int)VectorMode::RC>(
            cb_row_b, cb_row_b, Nt, false);
        // Keep the row sum: the softmax backward's row-sum correction divides by it, and it is
        // 1.0 only in exact arithmetic. Dropping it is worth up to 13.09x on ||dq||.
        copy_block(cb_row_b, cb_row_a, Nt);              // cb_row_a = rowsum(exp), cb_row_b free
        recip_block_inplace(cb_row_a, Nt);
        mul_block_bcast_cols<Nt, Nt, false, false>(cb_p, cb_row_a, cb_p);   // P = exp / rowsum
        cb_pop_front(cb_row_a, Nt);

        // ---- dV = P^T dO -------------------------------------------------------------------
        transpose_block<Nt, Nt>(cb_p, cb_t);
        mm_keep<Nt, Dt, Nt, col_sbh, col_sbw, false>(cb_t, cb_do, cb_dv);
        cb_pop_front(cb_t, score_tiles);

        // ---- dP = dO V^T, then dS = P * (dP - rowsum(dP * P) / rowsum(P)) --------------------
        mm_keep<Nt, Nt, Dt, sq_sbh, sq_sbw, true>(cb_do, cb_v, cb_dp);      // cb_dp = dP
        mul_block_to(cb_dp, cb_p, cb_t, score_tiles);    // cb_t = dP * P, both inputs kept
        reduce_c<PoolType::SUM, ReduceDim::REDUCE_ROW, cb_t, cb_scalar, Nt, (int)VectorMode::RC>(
            cb_row_a, cb_row_a, Nt, false);              // cb_row_a = rowsum(dP * P)
        cb_pop_front(cb_t, score_tiles);
        // P here is already normalised, so rowsum(P) is 1 up to bf16 rounding and this division is
        // exactly the correction that removes that rounding rather than a no-op.
        recip_block_inplace(cb_row_b, Nt);
        mul_block_inplace(cb_row_a, cb_row_b, Nt);
        cb_pop_front(cb_row_b, Nt);

        sub_block_bcast_cols_inplace<Nt, Nt>(cb_dp, cb_row_a);
        cb_pop_front(cb_row_a, Nt);
        mul_block_inplace(cb_dp, cb_p, score_tiles);     // cb_dp = dS
        cb_pop_front(cb_p, score_tiles);

        // ---- dbias += dS, the only term that reduces across the leading axis ----------------
        add_block_inplace<false>(cb_dbias, cb_dp, score_tiles);

        // ---- dQ = (dS K) * scale, dK = (dS^T Q) * scale --------------------------------------
        mm_keep<Nt, Dt, Nt, col_sbh, col_sbw, false>(cb_dp, cb_k, cb_dq);
        mul_block_bcast_scalar_inplace<cb_scale, col_tiles>(cb_dq);

        transpose_block<Nt, Nt>(cb_dp, cb_t);
        cb_pop_front(cb_dp, score_tiles);
        mm_keep<Nt, Dt, Nt, col_sbh, col_sbw, false>(cb_t, cb_q, cb_dk);
        mul_block_bcast_scalar_inplace<cb_scale, col_tiles>(cb_dk);
        cb_pop_front(cb_t, score_tiles);

        cb_pop_front(cb_q, col_tiles);
        cb_pop_front(cb_k, col_tiles);
        cb_pop_front(cb_v, col_tiles);
        cb_pop_front(cb_do, col_tiles);
    }

    // The writer drains cb_dbias's L1 directly rather than through a second buffer, so this is the
    // handshake that says the accumulator is final. A copy would have cost 331.8 KB of L1, which
    // at the shipped shape is the difference between fitting and not.
    cb_reserve_back(cb_done, 1);
    cb_push_back(cb_done, 1);
}

File: tt_bio/triatt_bw.py
"""Triangle attention's backward, with the score block never leaving L1.

`autograd.triangle_attention`'s backward recomputes one score block per chunk and frees it, which
keeps the 191.1 MB score tensor out of DRAM as a *resident* object but still moves it through DRAM
roughly 48 times per call: the matmul writes it, the scale reads and writes it, the bias add reads
and writes it, the softmax reads and writes it, and then the whole softmax backward reads and
writes it again. Measured at the shipped BindCraft 2 shape by `bcx-p10-bytes`, that is 962.6 GB a
round, 49.4 % of everything the round moves, on an op family already running at 84.5 % of the
card's DRAM roof. There is no per-byte speed left to find there; the only lever is fewer bytes.

The shape is what makes fewer bytes possible. Triangle attention's leading axis is the residue
axis itself, so a `[S, S, c]` pair tensor is S independent attention problems of length S, and at
the shipped 288-token axis one problem is

    q = k = v = dO   [288, 32] bf16    18.4 KB        scores   [288, 288] bf16   165.9 KB

with four heads. **N/d = 9**, so the scores are nine times q, and one whole problem plus its
gradients fits in a single core's L1 with room left over. A core here takes one head and a
contiguous group of the leading axis, forms S, P and dS in L1, and never writes any of them. What
crosses DRAM is q, k, v, dO in and dq, dk, dv out, which is 9x less than the score tensor alone.

THE BIAS GRADIENT IS WHY THIS CANNOT BE TWO OPS. The bias is `[1, H, S, S]` broadcast over the
leading axis, so `dbias` is the score gradient summed over that axis -- and dS exists only inside
this kernel. Splitting the gradient computation from the bias reduction would mean materialising
dS, which is the thing being avoided. So each core carries a `[S, S]` float32 accumulator in its
own L1 for its own group of the leading axis, writes it once at the end, and a single `ttnn.sum`
over the group axis finishes it. Float32 and not bf16: the accumulator sums a few hundred terms
and the reduce sums the partials on top of that, and a bf16 accumulator would put a rounding floor
under the one gradient this kernel produces by reduction rather than by matmul.

The key axis is never chunked, which is what keeps log-sum-exp bookkeeping out of this file. A
flash kernel chunks keys, so every block sees a partial softmax denominator and has to carry
running row statistics and rescale. Here each core holds every key for its rows, so each softmax
is exact and complete the first time. The QUERY axis may be chunked freely -- `dbias` rows are
indexed by the query row, so a query chunk needs no cross-chunk reduction -- and that is the knob
that carries this to longer token axes, where `bias + P + dS` stops fitting whole.

Driven through `tt_bio.sdpa_generic`'s machinery rather than the wheel's SDPA kernels: the forward
inherits `reader_interleaved.cpp`'s chain-forwarding, multicast, paging and MLA arguments, all of
which are dead at this call, and a backward built on top of them would be harder to read than
three small kernels of its own. What is inherited is the part that earned its place -- the CB
table, the work split, and the forward's own proof that a head's bias can sit permanently fronted
in L1 while the batch streams past it.

Default OFF behind `TT_BIO_TRIATT_BW_FUSED`. Gated in `tt_bio/autograd.py::triangle_attention`,
which keeps its chunked-recompute backward as the fallback for every shape this refuses.
"""

from __future__ import annotations

import math
import os
from functools import lru_cache
from pathlib import Path

import ttnn

from . import sdpa_generic as SG
from .envflags import env_flag, env_int

TILE = 32

# The whole-query form needs `bias`, `P`, `dS` and a transpose scratch live at once, plus a float32
# `dbias` accumulator that does not shrink with the query chunk. Past this many tiles on the token
# axis the whole-query form stops fitting and the caller falls back; `q_chunk_tiles` is what pulls
# it back under the line at a longer axis.
# Blackhole has 1.5 MB of L1 per core; tt-metal reserves the bottom of it for the program itself,
# and `sdpa_generic` already carries the figure the forward was sized against.
L1_PER_CORE = getattr(SG, "L1_PER_CORE", 1499136)
PROGRAM_RESERVE = getattr(SG, "PROGRAM_RESERVE", 0)

FUSED = env_flag("TT_BIO_TRIATT_BW_FUSED", False)

STATS = {"served": 0, "declined": 0}


def _kdir() -> Path:
    return Path(__file__).parent / "kernels" / "triatt_bw"


def _div_up(a: int, b: int) -> int:
    return (a + b - 1) // b


def _f32_bits(x: float) -> int:
    import struct
    return struct.unpack("<I", struct.pack("<f", float(x)))[0]


def plan(B: int, H: int, N: int, d: int, grid, q_chunk_tiles=None, dst_size=8):
    """Work split and tile geometry. Card-free, so the gate can be tested without a device.

    One core owns one head and a contiguous group of the leading axis. Heads go to the FAST index
    so that the cores sharing a head are adjacent, which is what lets the bias for that head be
    read once per core rather than once per leading-axis row.
    """
    if N % TILE or d % TILE:
        raise ValueError(f"token axis {N} and head dim {d} must both be tile aligned")
    Nt, Dt = N // TILE, d // TILE
    Qt = Nt if q_chunk_tiles is None else min(int(q_chunk_tiles), Nt)
    while Qt > 1 and Nt % Qt:
        Qt -= 1

    gx, gy = grid
    # Never more groups than there are leading-axis rows: a core with no row still allocates its
    # dbias accumulator and still contributes a partial to the reduce, so an empty one is pure
    # cost in both L1 and DRAM.
    groups = max(1, min(B, (gx * gy) // H))
    num_cores = groups * H
    rows_per_core = _div_up(B, groups)

    p = {
        "B": B, "H": H, "N": N, "d": d, "Nt": Nt, "Dt": Dt, "Qt": Qt,
        "groups": groups, "num_cores": num_cores, "rows_per_core": rows_per_core,
        "gx": gx, "gy": gy, "dst_size": dst_size,
    }
    # Priced off the CB table itself rather than a parallel formula. The two drifting apart is how
    # a gate ends up reporting that a config fits while tt-metal refuses it.
    p["l1_bytes"] = cb_bytes(p)
    return p


def fits_l1(p) -> bool:
    """Whether tt-metal will accept this config's CBs, priced the way `sdpa_generic` prices."""
    return p["l1_bytes"] + PROGRAM_RESERVE <= L1_PER_CORE


def largest_fitting_q_chunk(B, H, N, d, grid, **kw):
    """The biggest query chunk that fits, or None if even one tile row does not.

    The query axis is the only knob that shrinks the score-sized buffers, and shrinking it costs
    nothing in DRAM bytes or in arithmetic: `dbias` rows are indexed by the query row, so query
    chunks do not have to be reduced against each other. It does not touch the key axis, which is
    what would force log-sum-exp bookkeeping.
    """
    Nt = N // TILE
    for qt in range(Nt, 0, -1):
        if Nt % qt:
            continue
        p = plan(B, H, N, d, grid, q_chunk_tiles=qt, **kw)
        if fits_l1(p):
            return p
    return None


def eligible(q, k, v, bias, *, taping=False):
    """Whether this call can be served, with the reason recorded rather than guessed.

    Deliberately narrow. Everything it refuses falls through to the chunked-recompute backward,
    which is correct at every shape; there is no configuration where refusing costs a wrong answer.
    """
    if not FUSED:
        return False, "flag off"
    if bias is None:
        return False, "no bias: the chunked path is already cheap without a dbias reduction"
    qs = [int(x) for x in q.shape]
    if len(qs) != 4:
        return False, f"expected [B, H, N, d], got {qs}"
    B, H, N, d = qs
    if [int(x) for x in k.shape] != qs or [int(x) for x in v.shape] != qs:
        return False, "q, k and v must share a shape; cross attention is not this op"
    bs = [int(x) for x in bias.shape]
    if bs != [1, H, N, N]:
        return False, f"bias must be [1, {H}, {N}, {N}] broadcast over the batch, got {bs}"
    if N % TILE or d % TILE:
        return False, f"ragged: N={N} d={d} are not both tile aligned"
    if any(t.dtype != ttnn.bfloat16 for t in (q, k, v, bias)):
        return False, "bf16 operands only"
    return True, "ok"


# CB indices. Laid out so the three score-sized buffers are adjacent and easy to price.
CB_Q, CB_K, CB_V, CB_DO = 0, 1, 2, 3
CB_BIAS = 4                      # [Nt, Nt] for one head, fronted once and indexed, never popped
CB_SCALAR = 5                    # the packed bf16 1.0 the row reductions scale by
CB_P = 24                        # S, then P, in place
CB_DP = 25                       # dP, then dS, in place
CB_T = 26                        # transpose scratch: P^T for dV, dS^T for dK
CB_ROW_A, CB_ROW_B = 27, 28      # row max, then row sum, then its reciprocal
CB_DBIAS = 29                    # float32 accumulator, [Nt, Nt], persistent across the whole group
CB_DQ, CB_DK, CB_DV = 16, 17, 18

# There is deliberately no separate output buffer for `dbias`. The accumulator is read and written
# in place across the whole group and then pushed once, so the writer drains the same L1 the
# compute kernel accumulated into. A second [Nt, Nt] float32 buffer would have cost 331.8 KB to
# hold a copy of data that is already sitting there, and at the shipped shape that is the
# difference between fitting in L1 and not.


def cb_table(p):
    """(buffer index, tiles, page bytes, data format), priced the same way `sdpa_generic` prices."""
    Nt, Dt, Qt = p["Nt"], p["Dt"], p["Qt"]
    bf16, f32 = ttnn.bfloat16, ttnn.float32
    b16, b32 = 2048, 4096
    return [
        (CB_Q, Nt * Dt * 2, b16, bf16),
        (CB_K, Nt * Dt * 2, b16, bf16),
        (CB_V, Nt * Dt * 2, b16, bf16),
        (CB_DO, Nt * Dt * 2, b16, bf16),
        (CB_BIAS, Nt * Nt, b16, bf16),
        (CB_SCALAR, 1, b16, bf16),
        (CB_P, Qt * Nt, b16, bf16),
        (CB_DP, Qt * Nt, b16, bf16),
        (CB_T, Qt * Nt, b16, bf16),
        (CB_ROW_A, Qt, b16, bf16),
        (CB_ROW_B, Qt, b16, bf16),
        (CB_DBIAS, Nt * Nt, b32, f32),
        (CB_DQ, Nt * Dt * 2, b16, bf16),
        (CB_DK, Nt * Dt * 2, b16, bf16),
        (CB_DV, Nt * Dt * 2, b16, bf16),
    ]


def cb_bytes(p) -> int:
    return sum(n * page for _i, n, page, _f in cb_table(p))


def partial_shape(p):
    """The shape of the `dbias` partial tensor the kernel writes and `ttnn.sum` reduces.

    One `[H, N, N]` slab per group of the leading axis. `dim=0` of this is the only reduction the
    host does, and it is the whole reason a core can keep its accumulator in L1.
    """
    return [p["groups"], p["H"], p["N"], p["N"]]


def byte_census(p, elem=2, partial_elem=4):
    """What this program moves through DRAM, per invocation, in bytes.

    Same convention as the campaign's census: operand and result bytes on the padded device shape.
    Stated here rather than in a comment so an A/B can assert against it instead of a paragraph.
    """
    B, H, N, d = p["B"], p["H"], p["N"], p["d"]
    qkv = B * H * N * d * elem
    head_bias = H * N * N * elem
    per_core_bias = N * N * elem
    partial = p["groups"] * H * N * N * partial_elem
    return {
        "read_qkvdo": 4 * qkv,
        "read_bias": p["num_cores"] * per_core_bias,
        "write_dqdkdv": 3 * qkv,
        "write_dbias_partials": partial,
        "reduce_read_partials": partial,
        "reduce_write_dbias": head_bias,
        "total": 4 * qkv + p["num_cores"] * per_core_bias + 3 * qkv
                 + 2 * partial + head_bias,
    }


def chunked_recompute_bytes(p, passes=48, elem=2):
    """What the path this replaces moves, for the same invocation: `passes` over the score tensor.

    `passes` is measured, not assumed -- `bcx-p10-bytes` read 962.6 GB a round over 104
    invocations at the shipped shape, which is 48.4 score-tensor passes per call.
    """
    return int(passes * p["B"] * p["H"] * p["N"] * p["N"] * elem)
